File: 1 - LinkedList/C/3 - Circular Single Linked List/CircularSingleLinkedList.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Define macros to simplify memory allocation and deallocation
#define new(var) (var *)malloc(sizeof(var)) // Allocates memory for a new variable of type 'var'
#define delete(var) \
    free(var);      \
    var = NULL; // Frees allocated memory and sets the pointer to NULL to prevent dangling references

// Structure for a node
typedef struct Node {
    char id[11];
    char name[50];
    char major[30];
    struct Node* next;
} Node;

// Structure to manage the linked list
typedef struct LinkedList {
    Node* head; // Pointer to the head of the list
    Node* tail; // Pointer to the tail of the list
    int size;   // Size of the list
} LinkedList;

// Function prototypes
void createCircularSL(LinkedList* list, const char* id, const char* name, const char* major);
void addAtFirstNode(LinkedList* list, const char* id, const char* name, const char* major);
void addAtNthNode(LinkedList* list, const char* id, const char* name, const char* major, int pos);
void addAtLastNode(LinkedList* list, const char* id, const char* name, const char* major);
void updateFirstNode(LinkedList* list, const char* id, const char* name, const char* major);
void updateNthNode(LinkedList* list, const char* id, const char* name, const char* major, int pos);
void updateLastNode(LinkedList* list, const char* id, const char* name, const char* major);
void removeFirstNode(LinkedList* list);
void removeNthNode(LinkedList* list, int pos);
void removeLastNode(LinkedList* list);
void printCircularSL(LinkedList* list);
int countNodes(LinkedList* list);

int main() {
    // Initialize the linked list
    LinkedList list = { NULL, NULL, 0 };

    // Create the circular singly linked list with the first node
    createCircularSL(&list, "11223399", "Arief F-sa", "Computer Science");
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Add a node at the beginning of the list
    addAtFirstNode(&list, "11223398", "Nisa Uswa", "Psychology");
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Add a node at the end of the list
    addAtLastNode(&list, "11223400", "Pejeel", "Drop out");
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Remove the first node from the list
    removeFirstNode(&list);
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Remove the last node from the list
    removeLastNode(&list);
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Add nodes at the beginning and end of the list
    addAtFirstNode(&list, "11223398", "Nisa Uswa", "Psychology");
    addAtLastNode(&list, "11223400", "Pejeel", "Drop out");
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Add nodes at specific positions in the list
    addAtNthNode(&list, "11223401", "dummy1", "Math", 3);
    addAtNthNode(&list, "11223402", "dummy2", "Physics", 2);
    printCircularSL(&list);
    printf("==========\n\n\n");

    // Remove the node at the specified position in the list
    removeNthNode(&list, 2);
    printCircularSL(&list);
    printf("==========\n\n\n");

    return 0;
}

// ========== CREATE ==========
// Function to create a circular singly linked list with the first node
void createCircularSL(LinkedList* list, const char* id, const char* name, const char* major) {
    Node* newNode = new(Node);
    strcpy(newNode->id, id);
    strcpy(newNode->name, name);
    strcpy(newNode->major, major);
    newNode->next = newNode;
    list->head = newNode;
    list->tail = newNode;
    list->size = 1;
}

// ========== ADD ==========
// Function to add a node at the beginning of the list
void addAtFirstNode(LinkedList* list, const char* id, const char* name, const char* major) {
    if (list->size == 0) {
        createCircularSL(list, id, name, major);
        return;
    }
    Node* newNode = new(Node);
    strcpy(newNode->id, id);
    strcpy(newNode->name, name);
    strcpy(newNode->major, major);
    newNode->next = list->head;
    list->tail->next = newNode;
    list->head = newNode;
    list->size++;
}

// Function to add a node at the nth position in the list
void addAtNthNode(LinkedList* list, const char* id, const char* name, const char* major, int pos) {
    if (pos < 1 || pos > list->size + 1) {
        printf("Invalid position!\n");
        return;
    }
    if (pos == 1) {
        addAtFirstNode(list, id, name, major);
        return;
    }
    if (pos == list->size + 1) {
        addAtLastNode(list, id, name, major);
        return;
    }
    Node* newNode = new(Node);
    strcpy(newNode->id, id);
    strcpy(newNode->name, name);
    strcpy(newNode->major, major);
    Node* cur = list->head;
    for (int i = 1; i < pos - 1; i++) {
        cur = cur->next;
    }
    newNode->next = cur->next;
    cur->next = newNode;
    list->size++;
}

// Function to add a node at the end of the list
void addAtLastNode(LinkedList* list, const char* id, const char* name, const char* major) {
    if (list->size == 0) {
        createCircularSL(list, id, name, major);
        return;
    }
    Node* newNode = new(Node);
    strcpy(newNode->id, id);
    strcpy(newNode->name, name);
    strcpy(newNode->major, major);
    newNode->next = list->head;
    list->tail->next = newNode;
    list->tail = newNode;
    list->size++;
}

// ========== UPDATE ==========
// Function to update the first node in the list
void updateFirstNode(LinkedList* list, const char* id, const char* name, const char* major) {
    if (list->size == 0) {
        printf("List is empty!\n");
        return;
    }
    strcpy(list->head->id, id);
    strcpy(list->head->name, name);
    strcpy(list->head->major, major);
}

// Function to update the nth node in the list
void updateNthNode(LinkedList* list, const char* id, const char* name, const char* major, int pos) {
    if (pos < 1 || pos > list->size) {
        printf("Invalid position!\n");
        return;
    }
    Node* cur = list->head;
    for (int i = 1; i < pos; i++) {
        cur = cur->next;
    }
    strcpy(cur->id, id);
    strcpy(cur->name, name);
    strcpy(cur->major, major);
}

// Function to update the last node in the list
void updateLastNode(LinkedList* list, const char* id, const char* name, const char* major) {
    if (list->size == 0) {
        printf("List is empty!\n");
        return;
    }
    strcpy(list->tail->id, id);
    strcpy(list->tail->name, name);
    strcpy(list->tail->major, major);
}

// ========== DELETE ===========
// Function to remove the first node from the list
void removeFirstNode(LinkedList* list) {
    if (list->size == 0) {
        printf("List is empty!\n");
        return;
    }
    Node* del = list->head;
    if (list->size == 1) {
        list->head = NULL;
        list->tail = NULL;
    }
    else {
        list->head = list->head->next;
        list->tail->next = list->head;
    }
    delete(del);
    list->size--;
}

// Function to remove the nth node from the list
void removeNthNode(LinkedList* list, int pos) {
    if (pos < 1 || pos > list->size) {
        printf("Invalid position!\n");
        return;
    }
    if (pos == 1) {
        removeFirstNode(list);
        return;
    }
    Node* cur = list->head;
    for (int i = 1; i < pos - 1; i++) {
        cur = cur->next;
    }
    Node* del = cur->next;
    cur->next = del->next;
    if (del == list->tail) {
        list->tail = cur;
    }
    delete(del);
    list->size--;
}

// Function to remove the last node from the list
void removeLastNode(LinkedList* list) {
    if (list->size == 0) {
        printf("List is empty!\n");
        return;
    }
    Node* del = list->tail;
    if (list->size == 1) {
        list->head = NULL;
        list->tail = NULL;
    }
    else {
        Node* cur = list->head;
        while (cur->next != list->tail) {
            cur = cur->next;
        }
        list->tail = cur;
        list->tail->next = list->head;
    }
    delete(del);
    list->size--;
}

// ========== READ ===========
// Function to print the circular singly linked list
void printCircularSL(LinkedList* list) {
    if (list->size == 0) {
        printf("List is empty!\n");
        return;
    }
    printf("Number of Nodes: %d\n", countNodes(list)); // Display the total number of nodes

    Node* cur = list->head;
    int i = 1;
    do {
        printf("Node %d\n", i);
        printf("ID\t\t: %s\n", cur->id);
        printf("Name\t\t: %s\n", cur->name);
        printf("Major\t\t: %s\n\n\n", cur->major);
        cur = cur->next;
        i++;
    } while (cur != list->head);
}

// ========== UTILS ===========
// Function to count the number of nodes in the list
int countNodes(LinkedList* list) {
    return list->size;
}

File: 1 - LinkedList/Javascript/3-CircularSingleLinkedList/circularSingleLinkedList.js
/**
 * Class representing a node in the circular singly linked list.
 */
class Node {
    /**
     * Create a node.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    constructor(id, name, major) {
        this.id = id;
        this.name = name;
        this.major = major;
        this.next = null; // Pointer to the next node
    }
}

/**
 * Class representing a circular singly linked list.
 */
class LinkedList {
    /**
     * Create a linked list.
     */
    constructor() {
        this.head = null; // Pointer to the head of the list
        this.tail = null; // Pointer to the tail of the list
        this.size = 0; // Size of the list
    }

    /**
     * Create a circular singly linked list with the first node.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    createCircularSL(id, name, major) {
        const newNode = new Node(id, name, major);
        newNode.next = newNode;
        this.head = newNode;
        this.tail = newNode;
        this.size = 1;
    }

    /**
     * Add a node at the beginning of the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    addAtFirstNode(id, name, major) {
        if (this.size === 0) {
            this.createCircularSL(id, name, major);
            return;
        }
        const newNode = new Node(id, name, major);
        newNode.next = this.head;
        this.tail.next = newNode;
        this.head = newNode;
        this.size++;
    }

    /**
     * Add a node at the nth position in the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     * @param {number} pos - The position to insert the node.
     */
    addAtNthNode(id, name, major, pos) {
        if (pos < 1 || pos > this.size + 1) {
            console.log("Invalid position!");
            return;
        }
        if (pos === 1) {
            this.addAtFirstNode(id, name, major);
            return;
        }
        if (pos === this.size + 1) {
            this.addAtLastNode(id, name, major);
            return;
        }
        const newNode = new Node(id, name, major);
        let current = this.head;
        for (let i = 1; i < pos - 1; i++) {
            current = current.next;
        }
        newNode.next = current.next;
        current.next = newNode;
        this.size++;
    }

    /**
     * Add a node at the end of the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    addAtLastNode(id, name, major) {
        if (this.size === 0) {
            this.createCircularSL(id, name, major);
            return;
        }
        const newNode = new Node(id, name, major);
        newNode.next = this.head;
        this.tail.next = newNode;
        this.tail = newNode;
        this.size++;
    }

    /**
     * Update the first node in the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    updateFirstNode(id, name, major) {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        this.head.id = id;
        this.head.name = name;
        this.head.major = major;
    }

    /**
     * Update the nth node in the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     * @param {number} pos - The position of the node to update.
     */
    updateNthNode(id, name, major, pos) {
        if (pos < 1 || pos > this.size) {
            console.log("Invalid position!");
            return;
        }
        let current = this.head;
        for (let i = 1; i < pos; i++) {
            current = current.next;
        }
        current.id = id;
        current.name = name;
        current.major = major;
    }

    /**
     * Update the last node in the list.
     * @param {string} id - The ID of the data.
     * @param {string} name - The name of the data.
     * @param {string} major - The major associated with the data.
     */
    updateLastNode(id, name, major) {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        this.tail.id = id;
        this.tail.name = name;
        this.tail.major = major;
    }

    /**
     * Remove the first node from the list.
     */
    removeFirstNode() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        const del = this.head;
        if (this.size === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
            this.tail.next = this.head;
        }
        this.size--;
    }

    /**
     * Remove the nth node from the list.
     * @param {number} pos - The position of the node to remove.
     */
    removeNthNode(pos) {
        if (pos < 1 || pos > this.size) {
            console.log("Invalid position!");
            return;
        }
        if (pos === 1) {
            this.removeFirstNode();
            return;
        }
        let current = this.head;
        for (let i = 1; i < pos - 1; i++) {
            current = current.next;
        }
        const del = current.next;
        current.next = del.next;
        if (del === this.tail) {
            this.tail = current;
        }
        this.size--;
    }

    /**
     * Remove the last node from the list.
     */
    removeLastNode() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        const del = this.tail;
        if (this.size === 1) {
            this.head = null;
            this.tail = null;
        } else {
            let current = this.head;
            while (current.next !== this.tail) {
                current = current.next;
            }
            this.tail = current;
            this.tail.next = this.head;
        }
        this.size--;
    }

    /**
     * Print all nodes in the circular singly linked list.
     */
    printCircularSL() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        console.log(`Number of Nodes: ${this.size}`);
        let current = this.head;
        let i = 1;
        do {
            console.log(`Node ${i}`);
            console.log(`ID\t\t: ${current.id}`);
            console.log(`Name\t\t: ${current.name}`);
            console.log(`Major\t\t: ${current.major}\n`);
            current = current.next;
            i++;
        } while (current !== this.head);
    }

    /**
     * Utility function to count the number of nodes in the list.
     * @returns {number} The total number of nodes.
     */
    countNodes() {
        return this.size;
    }
}

// Example usage
const list = new LinkedList();
list.createCircularSL("11223399", "Arief F-sa", "Computer Science");
list.printCircularSL();
console.log("==========\n");

list.addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
list.printCircularSL();
console.log("==========\n");

list.addAtLastNode("11223400", "Pejeel", "Drop out");
list.printCircularSL();
console.log("==========\n");

list.removeFirstNode();
list.printCircularSL();
console.log("==========\n");

list.removeLastNode();
list.printCircularSL();
console.log("==========\n");

list.addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
list.addAtLastNode("11223400", "Pejeel", "Drop out");
list.printCircularSL();
console.log("==========\n");

list.addAtNthNode("11223401", "dummy1", "Math", 3);
list.addAtNthNode("11223402", "dummy2", "Physics", 2);
list.printCircularSL();
console.log("==========\n");

list.removeNthNode(2);
list.printCircularSL();
console.log("==========\n");